### PR TITLE
[ListTile] fixes how background color is applied to ListTile

### PR DESCRIPTION
**Pull request by AyushBherwani1998**
## Description

Fixes how the background color is applied to `ListTile`. The previous implementation which I did for background color also didn't respect the `InkResponse`. Found it when I was doing the changes for PR, sorry :(

GIF for Current Implementation

## Related Issues

#63877

## Tests

I added the following tests:

- Update all the tests which used `ColoredBox` with `Ink` widget
- Added test to verify the shape is properly propagated to `Ink.decoration`.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by AyushBherwani1998** [comment]
@Hixie I thought the changes will not break anything, but apparently, the widgets which are using `ListTile` are breaking due to recent changes. I'll look into the failing tests to see if I could figure it out.

File: packages/flutter/test/material/checkbox_list_tile_test.dart [hunk 1]
```diff
@@ -131,7 +131,7 @@ void main() {
       )
     );
 
-    final Rect paddingRect = tester.getRect(find.byType(Padding));
+    final Rect paddingRect = tester.getRect(find.descendant(of: find.byType(SafeArea), matching: find.byType(Padding)));
```
Comment (AyushBherwani1998) [inline packages/flutter/test/material/checkbox_list_tile_test.dart:134]: `Ink` widget also uses `Padding` due to which this test was failing. I fixed this, but locally one more test "ListTile caches baseline" is failing in `baseline_test.dart`.

Comment (AyushBherwani1998) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:134]: Also, should I just wrap the `InkWell` with `Material` and provide `shape` and `color`?

Comment (AyushBherwani1998) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:134]: Also, are the customer test failing due to the limitations of `Ink` widget as mentioned in the documentation?

**Comment by Hixie** [comment]
(You probably also want some golden tests to make sure the actual rendering is what you want.)

File: packages/flutter/test/material/checkbox_list_tile_test.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/test/material/checkbox_list_tile_test.dart:134]: I would need to investigate the failing tests to figure out exactly what is causing the failures. It's not obvious from the logs. I recommend checking out each test in turn. For each one, you'll want to step through the test, maybe running it interactively using `flutter run`, maybe using the debugger, maybe using tools like `debugDumpApp` and `debugDumpRenderTree`, comparing how they work with and without your patch.

It's probably the case that there's just one problem that's causing all the failures, that's pretty common.

**Comment by AyushBherwani1998** [comment]
@Hixie I figured out why the "ListTile caches baseline" is failing in `baseline_test.dart`. The `Ink` widget uses `LayoutBuilder` which extends the `RenderBox` and does not implement the `computeDistanceToActualBaseline`. So when `Baseline` calls the `child.getDistanceToActualBaseline` it doesn't propagate through the `Ink` widget to `__RenderListTile`.  It was working before because when `Baseline` calls the `child.getDistanceToActualBaseline`, it used to call the `_RenderListTile.child.getDistanceToActualBaseline` -> `title.getDistanceToActualBaseline` -> [`RenderBaselineDetector.computeDistanceToActualBaseline`.](https://github.com/flutter/flutter/blob/3562b3c0de3baa990d61f852e6ff12e843223fc6/packages/flutter/test/widgets/baseline_test.dart#L133)

**Comment by Hixie** [comment]
Ooh, that's unfortunate... cc @goderbauer 

I can't think of a way to solve that, unless we can change Ink to not use a LayoutBuilder...

**Comment by AyushBherwani1998** [comment]
What if instead of wrapping the `InkWell` with `Ink` widget, I use `Material` widget? It has the `color` and `shape` property that we can use to provide the desired output.

**Comment by Hixie** [comment]
(Changing Ink to not use a LayoutBuilder might be the way to go -- giving it a specialized private RenderObject for whatever purpose we currently use the LayoutBuilder.)

Using Material would prevent splashes inside the ListTile from splashing outside the ListTile.

**Comment by AyushBherwani1998** [comment]
The main purpose of using `LayoutBuilder` I guess is to provide the `RenderBox` object to the `InkDecoration`.

**Comment by AyushBherwani1998** [comment]
@Hixie I have not familiar with Flutter's rendering pipeline and it's algorithm. I would like to explore that part, any leads how should I approach to provide Ink widget a private RenderObject. I'll look at the implementation of Flex and ListTIle which uses RenderBox and try to figure out.

**Comment by Hixie** [comment]
Looking for other code in the widgets/ or material/ directory that has code that extends one of the Render* classes is probably the best place to start. There's a bunch of documentation in the API docs around RenderBox, RenderProxyBox, RenderObjectWidget, etc, too.

**Comment by AyushBherwani1998** [comment]
@Hixie Sorry to bother you again. I did try to provide a private `SingleChildRenderObjectWidget` which creates a `RenderBox` and uses `RenderObjectWithChildMixin<RenderBox>` . When I'm calling `getDistanceToBaseline` method on child inside `performLayout` it works as we want, but how to get a baseline of the parent? Usually, we call the getDistanceToBasline we have access to baseline, but for Ink widget, it will not be the case.

**Comment by Hixie** [comment]
> Also, any specific reasons why we can't override `computeDistanceToActualBaseline` and call `child.getDistanceToActualBaseline` if the child is not null in `LayoutBuilder`? We will still return `super.computeDistanceToActualBaseline(baseline)` but will propagate the baseline to the child. I did try locally and all tests passed.

I'm not sure I understand the question. Do you have sample code showing what you'd like to do?

**Comment by AyushBherwani1998** [comment]
> I'm not sure I understand the question. Do you have sample code showing what you'd like to do?

Something like this, since LayoutBuilder uses `RenderBox` we can override `computeDistanceToActualBaseline`. We are not overriding it to have a different computation for the baseline, we are just passing it to the child by calling `getDistanceToActualBaseline`. I tried to add this code in the local copy of the SDK and it passes all the tests.

```dart
@override
  double computeDistanceToActualBaseline(TextBaseline baseline) {
    if(child != null) {
      return child.getDistanceToActualBaseline(baseline);
    }
    return super.computeDistanceToActualBaseline(baseline);
  }
```

**Comment by Hixie** [comment]
Ohh, you're saying why not just make LayoutBuilder implement the baseline logic! Got it. Good question...

**Comment by Hixie** [comment]
...I can see no reason why we should not do that. That makes a lot of sense and it is a mystery why we had not done it already!

**Comment by AyushBherwani1998** [comment]
So, I will create a new issue that Layout Builder should implement baseline logic, and submit a PR for that. Once the PR is merged, I'll merge the master to pull the changes.

**Comment by AyushBherwani1998** [comment]
I have a submitted a PR #65000 . Let me know your thoughts.

**Comment by AyushBherwani1998** [comment]
@Hixie The customer tests are failing due to some other reason. The baseline test in the framework are now getting passed. Can someone from the team look into customer testing?

**Comment by AyushBherwani1998** [comment]
@Hixie Gentle ping.

**Comment by talamaska** [comment]
@AyushBherwani1998 Would you consider exposing the tileColor parameter to SwitchListTile and CheckboxListTile?

**Comment by AyushBherwani1998** [comment]
@Hixie I found the source of error but haven't able to figure out why it's thrown. When ListTile is give to SliverPrototypeExtentList.prototypeItem it throws the exception. The same goes for Rainbow monkey [PublicCommsView](https://github.com/jocosocial/rainbowmonkey/blob/c808dee61cddd7245b17530b882eb48769bc9d74/lib/src/views/comms.dart#L394) and PrivateCommsView.

I was able to replicate the error here. If you run the below code with the proposed PR it will throw an error.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: CustomScrollView(
          slivers: <Widget>[
            SliverAppBar(
              title: Text("SliverAppBar Title"),
            ),
            SliverPrototypeExtentList(
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) => ListTile(
                  title: Text('hey'),
                ),
                childCount: 50,
              ),
              prototypeItem: ListTile(),
            ),
          ],
        ),
      ),
    );
  }
}
```

It's a framework assertion. Any tips on how to debug more.

**Comment by AyushBherwani1998** [comment]
> 
> 
> @AyushBherwani1998 Would you consider exposing the tileColor parameter to SwitchListTile and CheckboxListTile?

Hey @talamaska, I would love to contribute. If you can create an issue, that would be great so team can keep track. Once, this PR is merged, we can definitely work on that.

**Comment by AyushBherwani1998** [comment]
@Hixie I'll create a new issue to keep track, whenever we pass the `Ink` widget to `prototypeItem` it throws exceptions.

**Pull request closed without merging**